examples_R10/R10_static_server.c

```diff
@@ -0,0 +1,302 @@
+   * it forbids aceptable paths like "/this/one..here", but it doesn't
+   * do anything to prevent symlink following." */
+  if (strstr(decoded_path, ".."))
+    goto err;
+
+  char whole_path[PATH_MAX] = {0};
+  const char *type = NULL;
+  path_join(whole_path, static_dir, decoded_path);
+  char *real_file = realpath(whole_path, NULL);
+  if (real_file) {
+    strncpy(whole_path, real_file, sizeof(whole_path));
+    free(real_file);
+  } else {
+    /* check also if is there gz-ready static version */
+    type = guess_content_type(whole_path);
+
+    char gz_path[PATH_MAX + 3] = {0};
+    snprintf(gz_path, sizeof(gz_path), "%s.gz", whole_path);
+    char *real_file = realpath(gz_path, NULL);
+    if (real_file) {
+      evhttp_add_header(evhttp_request_get_output_headers(req),
+                        "Content-Encoding", "gzip");
+      strncpy(whole_path, real_file, sizeof(whole_path));
+      free(real_file);
+    } else {
+      fprintf(stderr, "File '%s' not found\n", whole_path);
+      evhttp_send_error(req, HTTP_NOTFOUND, NULL);
+      goto done;
+    }
+  }
+
+  if (stat(whole_path, &st) < 0) {
+    goto err;
+  }
+
+  if ((evb = evbuffer_new()) == NULL) {
+    evhttp_send_error(req, HTTP_INTERNAL, 0);
+    goto cleanup;
+  }
+
+  bool dir_mode = false;
+
+  if (S_ISDIR(st.st_mode)) {
+    /* Check if there is index.html file */
+    char index_file[PATH_MAX + 11];
+    snprintf(index_file, sizeof(index_file), "%s/index.html", whole_path);
+
+    if (stat(index_file, &st) < 0)
+      dir_mode = true;
+    else
+      strcpy(whole_path, index_file);
+  }
+
+  if (dir_mode) {
+    DIR *d;
+    struct dirent *ent;
+
+    const char *trailing_slash = "";
+
+    if (!strlen(path) || path[strlen(path) - 1] != '/')
+      trailing_slash = "/";
+    if (!(d = opendir(whole_path))) {
+      goto err;
+    }
+
+    evbuffer_add_printf(
+        evb,
+        "<!DOCTYPE html>\n"
+        "<html lang=\"en\">"
+        "<head>\n"
+        "<meta name=\"viewport\" "
+        "content=\"width=device-width,initial-scale=1\">\n"
+        "<title>%s</title>\n"
+        "<link rel=\"shortcut icon\" href=\"/favicon.png\">\n"
+        "<link rel=\"stylesheet\" href=\"" BOOTSTRAP_CSS
+        "/bootstrap.min.css\">\n"
+        "<script src=\"" BOOTSTRAP_JS "/bootstrap.bundle.min.js\"></script>\n"
+        "<base href='%s%s'>\n"
+        "</head>\n"
+        "<body id=\"top\">\n"
+        "<nav class=\"navbar navbar-expand-lg navbar-dark sticky-top\">\n"
+        "<div class=\"container\">\n"
+        "</div>\n"
+        "</nav>\n"
+        "<main>\n"
+        "<div class=\"container p-3\">\n"
+        "<h2>%s</h2>\n"
+        "<ul class=\"list-unstyled my-3\">\n",
+        decoded_path, /* XXX html-escape this. */
+        path,         /* XXX html-escape this? */
+        trailing_slash, decoded_path /* XXX html-escape this */);
+    while ((ent = readdir(d))) {
+      const char *name = ent->d_name;
+      if (strcmp(name, "."))
```

Comment:
No point in showing `..` since it is prohibited above anyway

Reply:
I've just remembered my idea: '..' should be shown to support go up in page hierarchy. Anyway I've added another condition to exclude it for root path